server/src/main/java/org/opensearch/search/resource_limit_group/tracker/ResourceLimitsGroupResourceUsageTrackerService.java

```diff
@@ -0,0 +1,300 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.search.resource_limit_group.tracker;
+
+import org.opensearch.cluster.metadata.ResourceLimitGroup;
+import org.opensearch.cluster.service.ClusterService;
+import org.opensearch.common.inject.Inject;
+import org.opensearch.core.tasks.resourcetracker.TaskResourceUsage;
+import org.opensearch.search.resource_limit_group.ResourceLimitGroupPruner;
+import org.opensearch.search.resource_limit_group.ResourceLimitGroupTask;
+import org.opensearch.search.resource_limit_group.cancellation.CancellableTaskSelector;
+import org.opensearch.search.resource_limit_group.cancellation.ResourceLimitGroupTaskCanceller;
+import org.opensearch.tasks.Task;
+import org.opensearch.tasks.TaskCancellation;
+import org.opensearch.tasks.TaskManager;
+import org.opensearch.tasks.TaskResourceTrackingService;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+
+/**
+ * This class tracks requests per resourceLimitGroups
+ */
+public class ResourceLimitsGroupResourceUsageTrackerService extends ResourceLimitGroupTaskCanceller
```

Comment:
This class is overburdened with multiple responsibilities - Tracking resource usage, Managing task cancellations and Handling resource pruning and deletion.
I think we need to break them into their own classes. 

Something like 

```java
public class ResourcePruningService implements ResourcePruner {}
public class TaskCancellationService implements TaskCanceller {}
public class ResourceUsageService implements ResourceUsageTracker {}
```

Reply:
Yes, We will be creating these concrete implementations. This PR is going to serve as the Parent PR for upcoming small PRs which are breakdown of this.